src/slstm_layer.cpp

```diff
@@ -0,0 +1,555 @@
+    }
+}
+
+void save_cov_hidden_cell_states_smoother(int time_step, int num_states,
+                                          std::vector<float> &var_c_prior,
+                                          std::vector<float> &mu_o_ga,
+                                          std::vector<float> &jcb_ca,
+                                          std::vector<float> &cov_hc)
+/*
+ */
+{
+    for (int i = 0; i < num_states; i++) {
+        // cov(h_{t},c_{t})
+        cov_hc[time_step * num_states + i] =
+            var_c_prior[i] * jcb_ca[i] * mu_o_ga[i];
+    }
+}
+
+void save_cov_hidden_states_smoother(
+    std::vector<float> &mw, std::vector<float> &Jf_ga,
+    std::vector<float> &mi_ga, std::vector<float> &Ji_ga,
+    std::vector<float> &mc_ga, std::vector<float> &Jc_ga,
+    std::vector<float> &mo_ga, std::vector<float> &Jo_ga,
+    std::vector<float> &var_h_prev, std::vector<float> &mc_prev,
+    std::vector<float> &mca, std::vector<float> &Jca, int w_pos_f, int w_pos_i,
+    int w_pos_c, int w_pos_o, int no, int ni, int start_idx, int end_idx,
+    std::vector<float> &cov_hh)
+/*
+ */
+{
+    float Czz_f, Czz_i, Czz_c, Czz_o;
+    int m;
+
+    for (int t = start_idx; t < end_idx; t++) {
+        for (int j = 0; j < no; j++) {
+            // Forget gate
+            Czz_f = var_h_prev[t] * Jca[j] * mo_ga[j] * Jf_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_f] * mc_prev[j];
+
+            // Input gate
+            Czz_i = var_h_prev[t] * Jca[j] * mo_ga[j] * Ji_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_i] * mc_ga[j];
+
+            // Cell state gate
+            Czz_c = var_h_prev[t] * Jca[j] * mo_ga[j] * Jc_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_c] * mi_ga[j];
+
+            // Output gate
+            Czz_o = var_h_prev[t] * Jo_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_o] * mca[j];
+
+            // Updating quantities
+            m = t * no + j;
+            cov_hh[m] = Czz_f + Czz_i + Czz_c + Czz_o;
+        }
+    }
+}
+
+void flatten2DVector(const std::vector<std::vector<float>> &vec2D,
+                     std::vector<float> &vec1D) {
+    // Reserve space in vec1D for efficiency
+    vec1D.reserve(vec2D.size() * vec2D[0].size());
+
+    for (const auto &row : vec2D) {
+        vec1D.insert(vec1D.end(), row.begin(), row.end());
+    }
+}
+
+void smooth_cell_states(
+    int num_timestep, int num_states, std::vector<float> &cov_cc,
+    std::vector<float> &mu_c_priors, std::vector<float> &var_c_priors,
+    std::vector<float> &mu_c_posts, std::vector<float> &var_c_posts,
+    std::vector<float> &mu_c_smooths, std::vector<float> &var_c_smooths)
+/*
+ */
+{
+    int current, next;
+    for (int i = num_timestep - 2; i >= 0; --i) {
+        for (int j = num_states - 1; j >= 0; --j) {
+            current = i * num_timestep + j;
+            next = (i + 1) * num_timestep + j;
+            float tmp = cov_cc[next] / var_c_priors[next];
```

Comment:
@van-dai-vuong the issue comes from here. In the c++ test, the size of `cov_cc` is 8640, but in an iteration, `next` index is 46479

Reply:
I fixed the bug and rebased again. I tested, it works on the server.